src/server/twemcache-rs/src/lib.rs

```diff
@@ -0,0 +1,322 @@
+                    error!("{}", e);
+                    std::process::exit(1);
+                }
+            }
+        } else {
+            Arc::new(Default::default())
+        };
+
+        if config.worker().threads() > 1 {
+            Self::multi(config)
+        } else {
+            Self::single(config)
+        }
+    }
+
+    fn multi(config: Arc<Config>) -> Self {
+        // initialize admin
+        let admin = Admin::new(config.clone()).unwrap_or_else(|e| {
+            error!("{}", e);
+            std::process::exit(1);
+        });
+
+        // initialize storage
+        let mut storage = Storage::new(config.clone()).unwrap_or_else(|e| {
+            error!("{}", e);
+            std::process::exit(1);
+        });
+
+        // initialize workers
+        let mut workers = Vec::new();
+        let mut session_senders = Vec::new();
+        for _ in 0..config.worker().threads() {
+            // let storage_channel = storage.add_queue();
+            let worker = MultiWorker::new(config.clone(), &mut storage).unwrap_or_else(|e| {
+                error!("{}", e);
+                std::process::exit(1);
+            });
+            session_senders.push(worker.session_sender());
+            workers.push(worker);
+        }
+
+        // initialize server
+        let server = Server::new(config, session_senders).unwrap_or_else(|e| {
+            error!("{}", e);
+            std::process::exit(1);
+        });
+
+        TwemcacheBuilder::MultiWorker(MultiWorkerTwemcacheBuilder {
+            admin,
+            server,
+            storage,
+            workers,
+        })
+    }
+
+    fn single(config: Arc<Config>) -> Self {
+        // initialize admin
+        let admin = Admin::new(config.clone()).unwrap_or_else(|e| {
+            error!("{}", e);
+            std::process::exit(1);
+        });
+
+        // initialize worker
+        let worker = SingleWorker::new(config.clone()).unwrap_or_else(|e| {
+            error!("{}", e);
+            std::process::exit(1);
+        });
+        let session_senders = vec![worker.session_sender()];
+
+        // initialize server
+        let server = Server::new(config, session_senders).unwrap_or_else(|e| {
+            error!("{}", e);
+            std::process::exit(1);
+        });
+
+        TwemcacheBuilder::SingleWorker(SingleWorkerTwemcacheBuilder {
+            admin,
+            server,
+            worker,
+        })
+    }
+
+    /// Converts the `TwemcacheBuilder` to a running `Twemcache` by spawning
+    /// the threads for each component. Returns a `Twemcache` which may be used
+    /// to block until the threads have exited or trigger a shutdown.
+    pub fn spawn(self) -> Twemcache {
+        match self {
+            TwemcacheBuilder::MultiWorker(b) => Self::spawn_multi(b),
+            TwemcacheBuilder::SingleWorker(b) => Self::spawn_single(b),
+        }
+    }
+
+    fn spawn_multi(builder: MultiWorkerTwemcacheBuilder) -> Twemcache {
```

Comment:
nitpick: since admin and server thread setup are the same between single and multi, can you order them such that related lines appear at the top of the function, followed by those related to worker & storage,  which are different between the two? This way you can add comments before each section saying "generic setup" and "setup specific to the multi/single-threading"